### PR TITLE
Add optional firefox container selection to --cookies-from-browser

### PR DESCRIPTION
**Pull request by bashonly**
</details>

This adds support for Firefox containers to the `--cookies-from-browser` option.

Containers are a relatively new [built-in Firefox feature](https://support.mozilla.org/en-US/kb/containers#w_for-advanced-users) that allow for cookie isolation. A user may have multiple containers that are logged in to the same site with different accounts. As such, using the `--cookies-from-browser firefox` option as it currently stands can cause issues with overlapping/conflicting auth cookies for users of Firefox containers, since the current behavior loads all cookies from FF's cookie database. This PR is intended to solve that problem.

The proposed new argument syntax is:
`--cookies-from-browser BROWSER[+KEYRING][:PROFILE[:CONTAINER]]`

The name of the Firefox container can be passed as optional argument `CONTAINER` that is appended to `PROFILE` with another `:` as a separator.
If the user wishes to specify a container, the `PROFILE` arg is then required.
Because `PROFILE` is allowed to be a path as well as a profile name, yt-dlp will check if the entire `[PROFILE[:CONTAINER]]` string is an existing path before splitting it into separate `profile` and `container` values.

yt-dlp then only loads cookies associated with that specifed container. 
If there are multiple Firefox containers with the same name, the first match is used. I thought this would be better than loading from all matching containers, since the problem that this improvement intends to solve is overlapping/conflicting cookies.
If no container is specified, the default behavior is to only load cookies that have no container.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [x] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/__init__.py [hunk 1]
```diff
@@ -352,12 +353,16 @@ def parse_chapters(name, value):
         if browser_name not in SUPPORTED_BROWSERS:
             raise ValueError(f'unsupported browser specified for cookies: "{browser_name}". '
                              f'Supported browsers are: {", ".join(sorted(SUPPORTED_BROWSERS))}')
+        elif browser_name == 'firefox':
+            if ':' in profile and not os.path.exists(profile):
+                container = ':'.join(profile.split(':')[1:])
+                profile = profile.split(':')[0]
```
Comment (pukkandan) [inline yt_dlp/__init__.py:359]:
```suggestion
                profile, container = profile.split(':', 1)
```

File: yt_dlp/cookies.py [hunk 2]
```diff
@@ -126,6 +128,24 @@ def _extract_firefox_cookies(profile, logger):
...
+        with open(containers_path, 'r') as containers:
+            identities = json.load(containers).get('identities', [])
+        for context in identities:
+            if context.get('name') == container:
+                container_id = context.get('userContextId')
+                break
+            elif context.get('l10nID'):
+                if re.match(r'userContext([^\.]+)\.label', context.get('l10nID')).group() == container:
+                    container_id = context.get('userContextId')
+                    break
```
Comment (pukkandan) [inline yt_dlp/cookies.py:145]:
```suggestion
        container_id = next((context.get('userContextId') for context in identities if container in (
            context.get('name'),
            try_call(lambda: re.fullmatch(r'userContext([^\.]+)\.label', context['l10nID']).group())
        )), None)
```
OR
```suggestion
        container_id = traverse_obj(identities, (
            lambda x: (container == x.get('name')
                       or container == re.fullmatch(r'userContext([^\.]+)\.label', x['l10nID']).group()),
            'userContextId',
        ), get_all=False)
```

File: yt_dlp/cookies.py [hunk 3]
```diff
@@ -135,7 +155,13 @@ def _extract_firefox_cookies(profile, logger):
         cursor = None
         try:
             cursor = _open_database_copy(cookie_database_path, tmpdir)
-            cursor.execute('SELECT host, name, value, path, expiry, isSecure FROM moz_cookies')
+            if isinstance(container_id, int):
```
Comment (pukkandan) [inline yt_dlp/cookies.py:158]: Don't check for int here and None above. Use the same check in both places

File: yt_dlp/cookies.py [hunk 4]
```diff
@@ -135,7 +155,13 @@ def _extract_firefox_cookies(profile, logger):
...
+                logger.debug(
+                    f'Only loading cookies from firefox container "{container}", ID {container_id}')
+                cursor.execute(
+                    f'SELECT host, name, value, path, expiry, isSecure FROM moz_cookies WHERE originAttributes="^userContextId={container_id}"')
+            else:
+                cursor.execute('SELECT host, name, value, path, expiry, isSecure FROM moz_cookies')
```
Comment (pukkandan) [inline yt_dlp/cookies.py:164]: In this case, load only cookies from "no container"

File: yt_dlp/options.py [hunk 5]
```diff
@@ -1399,15 +1399,17 @@ def _alias_callback(option, opt_str, value, parser, opts, nargs):
...
             'The name of the browser and (optionally) the name/path of '
             'the profile to load cookies from, separated by a ":". '
             f'Currently supported browsers are: {", ".join(sorted(SUPPORTED_BROWSERS))}. '
             'By default, the most recently accessed profile is used. '
             'The keyring used for decrypting Chromium cookies on Linux can be '
             '(optionally) specified after the browser name separated by a "+". '
-            f'Currently supported keyrings are: {", ".join(map(str.lower, sorted(SUPPORTED_KEYRINGS)))}'))
+            f'Currently supported keyrings are: {", ".join(map(str.lower, sorted(SUPPORTED_KEYRINGS)))}'
+            'Optionally, a Firefox container name can be specified after the '
+            'profile (which must then be specified), separated by a ":".'))
```
Comment (pukkandan) [inline yt_dlp/options.py:1412]:
```suggestion
            'profile, separated by a ":".'))
```
that is implied

**Review by pukkandan — CHANGES_REQUESTED** [review]
Update documentation in https://github.com/yt-dlp/yt-dlp/blob/e5458d1d88fcc81011ab19ba610c4b37946c9fa9/yt_dlp/YoutubeDL.py#L305-L307

File: yt_dlp/__init__.py (hunk 1 above)
Comment (bashonly) [reply yt_dlp/__init__.py]: done

File: yt_dlp/cookies.py (hunk 2 above)
Comment (bashonly) [reply yt_dlp/cookies.py]: done

File: yt_dlp/options.py (hunk 5 above)
Comment (bashonly) [reply yt_dlp/options.py]: done

File: yt_dlp/cookies.py (hunk 4 above)
Comment (bashonly) [reply yt_dlp/cookies.py]: done

File: yt_dlp/cookies.py (hunk 3 above)
Comment (bashonly) [reply yt_dlp/cookies.py:155]: done

File: yt_dlp/options.py [hunk 6]
```diff
@@ -1399,15 +1399,17 @@ def _alias_callback(option, opt_str, value, parser, opts, nargs):
...
             'The name of the browser and (optionally) the name/path of '
             'the profile to load cookies from, separated by a ":". '
             f'Currently supported browsers are: {", ".join(sorted(SUPPORTED_BROWSERS))}. '
             'By default, the most recently accessed profile is used. '
             'The keyring used for decrypting Chromium cookies on Linux can be '
             '(optionally) specified after the browser name separated by a "+". '
-            f'Currently supported keyrings are: {", ".join(map(str.lower, sorted(SUPPORTED_KEYRINGS)))}'))
+            f'Currently supported keyrings are: {", ".join(map(str.lower, sorted(SUPPORTED_KEYRINGS)))}'
+            'Optionally, a Firefox container name can be specified after the '
+            'profile separated by a ":".'))
```
Comment (pukkandan) [inline yt_dlp/options.py:1412]:
```suggestion
            'Optionally, a Firefox container name can be specified after the profile separated by a ":"'))
```

**Pull request merged**